### ISSUE TITLE
Onnx model with initializers as extra input

### ISSUE THREAD

**Issue by taoisu** [triton-inference-server/server#5489]
**Description**

I have an onnx model with a list of initializers which are tensors with default value but can be overwritten at runtime by input tensors with the same name as the tensor, because of that, the initializers are not in the model's input names, so when I tried to write the config for this model, I had to leave the initializers names out, otherwise I'll get an error. During inference, what I was doing with onnxruntime is to send input tensors along with the initializer tensors as the final input, but when I tries to do the same thing in triton it failed saying it wasn't expecting the extra input tensors. Any advice on how to make this work?

**Triton Information**
What version of Triton are you using?

22.09-py3

Are you using the Triton container or did you build it yourself?

container

**To Reproduce**
Steps to reproduce the behavior.

Describe the models (framework, inputs, outputs), ideally include the model configuration file (if using an ensemble include the model configuration file for that as well).

**Expected behavior**

**Comment by taoisu** [comment]
It looks like the initializers are not supposed to be present in the input list of onnxruntime, but it does present in the onnx model. During inference, when provided with more input tensors, the check fails here: https://github.com/triton-inference-server/core/blob/f59101d8449b1305f6f10c9bf39e182039a62f0f/src/infer_request.cc#:~:text=if%20((original_inputs_.size()%20%3E%20(size_t)model_config.input_size())%20%7C%7C

**Comment by taoisu** [comment]
Made a dummy model in case anyone needs a repro:
![image](https://user-images.githubusercontent.com/30015018/225173677-9de8ab8b-fb5b-4f27-8dba-4853160db777.png)

both hidden.weight and hidden.bias are made input of the model.

[dummy.zip](https://github.com/triton-inference-server/server/files/10974954/dummy.zip)

**Comment by taoisu** [comment]
@Tabrizian @GuanLuo saw you guys are active in this repo, do you have any suggestions? I'm actively checking the code, it seems to me there are assumptions made that inputs has the first dimension being the batch dimension, but it's not the case for initializers which are essentially just model weights. Any thoughts if there is an easy workaround or advice on how we want to make the change? Happy to contribute to the repo.

**Comment by taoisu** [comment]
Hi folks, after some trial here is my current thinking:
1. The initializers are essentially model weights, it's different from typical input for:
   1) The weights don't have batch dimension.
   2) Inputs with weights cannot use dynamic batching because you cannot batch weights with different values.

2. To support this, I'm thinking to:
    1) Add `is_initializer_tensor` property to `ModelInput` message in https://github.com/triton-inference-server/common/blob/main/protobuf/model_config.proto, make it only available for onnxruntime backend for the time being.
    2) During onnxruntime session creation, we skip passing initializer input
    3) Add special treatment to initializer tensor in places like `InferenceRequest::Normalize()` & `ModelInstanceState::SetInputTensors` to handle batch correctly
    4) Make sure dynamic batching is disabled when initializer tensor exists.

Any suggestions or feedbacks? @Tabrizian @GuanLuo @dyastremsky @oandreeva-nv (Apologies for mentioning many folks here, need someone's help on this topic. Thanks!)

**Comment by GuanLuo** [comment]
Just read through the ORT C APIs and I couldn't find the way to override initializer value dynamically, all related APIs I have seen is to be supplied when creating the ORT section. Given that, I don't think it makes much sense to allow it to be Triton request input if the underlying framework can't consume it at per request basis.

However, based on the ONNX definition, it appears that an [initializer can be specified as model input](https://github.com/onnx/onnx/blob/main/onnx/onnx.proto#L452), have you set that when exporting the model? Does ORT not treat it as regular model input in this case?

**Comment by taoisu** [comment]
@GuanLuo Thank you for your response!

Just as you described, our production model is exported by specifying some of the weights (initializers) as input and we use onnxruntime for inference. During inference, different request may provide different weights to the model and onnxruntime does support it well.

Some more info about initializer as model input can also be found here: https://github.com/onnx/onnx/blob/main/docs/IR.md#nodes

However, even after specifying initializers as input, the onnxruntime won't list those initializers as model input, only non-initializers input will be returned upon query, code can be found here: https://github.com/microsoft/onnxruntime/blob/2023836c2f97dd6fe417e0b6f2295efdd3346fb8/include/onnxruntime/core/graph/graph.h#L770

We have confirmation from onnxruntime team this is a supported behavior and we've been running this model in production for a couple of months now. Recently we're seriously considering migrating our workload to Triton that's when I hit this issue.

**Comment by taoisu** [comment]
What I can do is to provide a python inference code to you to verify ort is working in this case.

I've been playing with the Triton code a little bit more and have a revised version of the proposal:

1. Add a property `extra_input_as_initializers` in model_config.proto and make it only applicable to onnxruntime backend
2. When that property is on, the triton will pass the extra tensors as is to the onnxruntime backend.

The reason being sometimes there could be hundreds of initializers, this way the config can be more concise.

If we can agree on the proposals, I'm happy to work on the PRs in the next few days.

**Comment by taoisu** [comment]
Hi @GuanLuo, the following is a simple way to verifying onnxruntime's support for overriding model weights:

```python
import torch
import torch.nn as nn


class SimpleNet(nn.Module):

    def __init__(self):
        super(SimpleNet, self).__init__()
        self.hidden = nn.Parameter(torch.randn(4))

    def forward(self, x: torch.Tensor):
        x = self.hidden + x
        return x


torch.onnx.export(
    SimpleNet(),
    torch.randn(4),
    'dummy.onnx',
    keep_initializers_as_inputs=True,
    input_names=['in'],
    output_names=['out'],
    dynamic_axes={'in': {0: 'bs'}, 'out': {0: 'bs'}},
)

import onnxruntime as ort

sess = ort.InferenceSession('dummy.onnx', providers=['CPUExecutionProvider'])
print(sess.run(None, {'in': torch.zeros(4).numpy()}))
print(sess.run(None, {'in': torch.ones(4).numpy(), 'hidden': torch.arange(4, dtype=torch.float32).numpy()}))
```

Which outputs two lines, the first line is random, while the second line is deterministically `[1., 2., 3., 4.]`

**Comment by GuanLuo** [comment]
Thanks for your update, given your finding, I would imagine that ORT C API should also be able to handle accepting initializer for inputs if the model is being exported accordingly.

I think your proposal makes sense, though I have a few suggestions on top of it:
1. Instead of defining another config field, you may use `optional` to denote the input should correspond to an overridable initializer in the context of ORT backend.
2. Then you can use the input if provided in the way you described. You probably need to update the validate function in ORT backend so it looks up the tensor via different API ([`SessionGetOverridableInitializerTypeInfo`](https://github.com/microsoft/onnxruntime/blob/2023836c2f97dd6fe417e0b6f2295efdd3346fb8/include/onnxruntime/core/session/onnxruntime_c_api.h#L1077) ?)

I think all the changes will be contained in ORT backend, although like you mentioned, Triton's dynamic batching feature may not be usable as the "initializer" usually doesn't follow the batching assumption (1st dimension is batch dimension).

**Comment by taoisu** [comment]
Thanks @GuanLuo ! This is very helpful information! I was thinking of adding this new config field because in our case there are over 100 initializers to override, which would make the config.pbtxt very long. Do you feel strongly against this idea? If so, I can try the `optional` path, I still hope we can simplify the config file as much as possible.

**Comment by taoisu** [comment]
@GuanLuo Please take a look for the above two PRs when you have time. Thanks very much for your help!

**Comment by taoisu** [comment]
Hi @GuanLuo, just had a discussion with our legal colleagues, they have some concerns about the scope of the CLA, I asked them if they think it's ok that the PR is made by an NVIDIA engineer, and they told me it would be preferred that way. Do you think it's OK that you create a PR on this matter and I drop mine? I hope we can merge this PR sooner.

**Comment by taoisu** [comment]
Thanks @GuanLuo for the help!
